File: typescript/12_sorts/MergeSort.ts
/**
 * 归并排序
 * 稳定排序，稳定的O(nlgn)时间复杂度
 * O(n)的空间复杂度
 * 在小规模数据排序中很常用
 */
class MergeSort {
  public static mergeSort(array: number[]) {
    if (!array || !array.length) return
    const length = array.length
    this.mergeSortInternally(array, 0, length - 1)
  }

  static mergeSortInternally(array: number[], p: number, r: number) {
    if (p >= r) return
    // 严格按照中间值作切分点
    // js中除法需要做取整操作，不然结果有可能是小数
    const q = Math.floor(p + (r - p) / 2)
    this.mergeSortInternally(array, p, q)
    this.mergeSortInternally(array, q + 1, r)
    this.mergeArray(array, p, q, r)
  }

  private static mergeArray(a: number[], p: number, q: number, r: number) {
    let i = p
    let j = q + 1
    let k = 0
    // 定义一个临时数组来存放排序的值
    const tmp: number[] = []
    while (i <= q && j <= r) {
      if (a[i] <= a[j]) {
        tmp[k++] = a[i++]
      } else {
        tmp[k++] = a[j++]
      }
    }
    // 判断哪个子数组中有剩余的数据
    let start = i
    let end = q
    if (j <= r) {
      start = j
      end = r
    }
    // 将剩余的数据拷贝到临时数组tmp
    while (start <= end) {
      tmp[k++] = a[start++]
    }

    // 将tmp中的数组拷贝回a[p...r]
    for (i = 0; i <= r - p; i++) {
      a[p + i] = tmp[i]
    }
  }
}

const test4 = [1, 3, 2, 3, 10, 9, 7, 6, 0, 12]
MergeSort.mergeSort(test4)
console.log(test4)

File: rust/12_sorts/merge_sort.rs
/// 归并排序
/// 时间复杂度 O(nlogn), 空间复杂度：O(n), 稳定排序
// 归并排序
pub fn merge_sort(mut nums: Vec<i32>) -> Vec<i32> {
    if nums.is_empty() { return nums; }

    let n = nums.len() - 1;
    merge_sort_internally(&mut nums, 0, n);
    nums
}

fn merge_sort_internally(nums: &mut Vec<i32>, start: usize, end: usize) {
    if start >= end { return; }

    let middle = start + (end - start) / 2;
    merge_sort_internally(nums, start, middle);
    merge_sort_internally(nums, middle+1, end);

    // merge two array
    merge(nums, start, middle, end);
}

fn merge(nums: &mut Vec<i32>, start: usize, middle: usize, end: usize) {
    let mut i = start;
    let mut j = middle + 1;
    let mut tmp = vec![];
    while i <= middle && j <= end {
        if nums[i] <= nums[j] {
            tmp.push(nums[i]);
            i += 1;
        } else {
            tmp.push(nums[j]);
            j += 1;
        }
    }

    let mut s = i;
    let mut r = middle;
    if j <= end {
        s = j;
        r = end;
    }

    while s <= r {
        tmp.push(nums[s]);
        s += 1;
    }

    for i in 0..=(end-start) {
        nums[start+i] = tmp[i];
    }
}

fn main() {
    let nums = vec![8, 10, 2, 3, 6,1, 5];
    println!("{:?}", merge_sort(nums));
}
